keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,448 @@
+
+* `kube_running_pod_resource_requests` contains series for the consumed requested resources for a pod at a given time.
+* `kube_running_pod_resource_limits` contains series for the consumed limit resources for a pod at a given time.
+
+
+This will add `O(pods * resource types)` metrics series, and in general most clusters have 4-8 resource types. The series per pod is `non_zero_resource_count * 2 * 2 (requests and limits for normal and _lifecycle)`.  This explicitly does not expose containers as a dimension, as the resource model applies rules to containers that cannot be trivially expressed via a formula and does not help answer the core questions.
+
+These metrics should roughly correspond to:
+
+* The resources the scheduler will consider the pod to be consuming at the current time
+* The resources the kubelet will use to decide whether to admit the pod (nodes will reject requested resources that exceed their actual allocatable capacity)
+* The effective reserved capacity as resource quota would observe for the namespace
+* The sum of all resources consumed on the cluster
+* The size of cgroup constraints applied to pods on a node
+
+The lifecycle metrics allow key insights that are impossible today:
+
+* Observing when init containers or containers have substantially different resource reservation, causing workloads to under provision nodes
+* Visualize batch compute consumption on the cluster where completed pods can be quickly excluded
+
+The implementation would be disablable via a flag in the kube-scheduler `--disable-pod-resource-metrics` to allow distributions to opt-out to the increased cardinality, although in practice we expect the impact in scheduler to be drowned by other O(pods) cardinality in other parts of the system and so this flag should not be necessary.
+
+The metrics would be served by the scheduler that is holding the active scheduler lease. During a failover, no metrics would be reported. This would prevent duplicate metrics from being scraped and aligns with our decision in the scheduler to not fill caches prior to election for resource usage minimization.
+
+Custom schedulers based on the scheduler code should be able to easily reuse and report these metrics, and the configuration of the cluster instrumentation to retrieve those metrics would be left as an exercise for the integrator.
+
+
+### Add recording rules consistent with this metric to describe actual resource usage
+
+In order to make these metrics useful for calculation of actual usage, we would represent the current usage in a form that can easily be queried by applying recording rules in the standard Prometheus stack and equivalents in other systems.
+
+The relevant metric for pod CPU would be:
+
+* `kube_pod_resource_usage_cpu_seconds` - the cpu currently consumed by a running pod since last restart
```

Comment:
Since this is in core-seconds, it is a cumulative metric.  It should have the `_total` suffix.

Reply:
Simplified this.